File: 01. DataStructures/03. Queue/Basic Operations/05. Circular Queue using Linked List/circular_queue_linked_list.js
// - Enqueue() – Add item to the queue from the REAR.
// - Dequeue() – Remove item from the queue from the FRONT.
// - isFull() – Check if queue is full or not.
// - isEmpty() – Check if queue empty or not.
// - count() – Get number of items in the queue.
// - peek() – return front elemenet in the queue(line).
// 1->2->3->4

class Node {
    constructor(val, next = null) {
        this.val = val;
        this.next = next;
    }
}

class Circular_queue_linked_list {
    constructor(size) {
        this.size = size;
        this.counter = 0;
        this.queue = this.front;

        // null->null->null->null
        this.front = new Node(null);
        let temp = this.front;
        let i = 1;
        while(i < this.size) {
            const temp1 = new Node(null);
            this.front.next = temp1;
            this.front = this.front.next;
            i++;
        }
        this.front.next = temp;
        this.front = temp;
        this.rear = temp;
    }
    //1->2->3->4
    enqueue(data) {
        
        if(this.isFull()) {
            return;
        }else {
            this.rear.val = data;
            this.rear = this.rear.next;
            this.counter++;
        }
    }
    dequeu() {
        if(this.front.val) {
            this.front.val = null;
            this.front = this.front.next;   
            this.counter--; 
        }
    }
    empty() {
        if(this.front.val == null && this.rear.val == null) {
            return true;
        } else {
            return false;
        }
    }
    isFull() {
        if(this.rear.val == null) {
            return false;
        } else {
            return true;
        }
    }
    peek() {
        if(this.front.val) {
            return this.front.val;
        } else {
            return 'queue is empty.'
        }
    }
    count() {
        return this.counter;
    }
    print() {
        let output = '';
        let currunt = this.front;
        while(currunt.next !== this.front) {
            output +=  currunt.val + '->';
            currunt = currunt.next;
        }
        output += currunt.val + '->null';
        return output;
    }
}
const myCircularQueue = new Circular_queue_linked_list(4);
myCircularQueue.enqueue(1);
myCircularQueue.enqueue(2);
myCircularQueue.enqueue(3);
// myCircularQueue.dequeu();
// myCircularQueue.dequeu();
// myCircularQueue.dequeu();
// myCircularQueue.dequeu();
// myCircularQueue.dequeu();
// myCircularQueue.enqueue(4);
// myCircularQueue.dequeu();
// myCircularQueue.enqueue(5);
// myCircularQueue.dequeu();
// myCircularQueue.enqueue(10);
myCircularQueue.dequeu();
myCircularQueue.enqueue(2);
console.log(myCircularQueue.count());

File: 01. DataStructures/03. Queue/Basic Operations/05. Circular Queue using Linked List/CircularQueue_LinkedList.cpp
// C++ program for insertion and deletion in Circular Queue

#include <bits/stdc++.h>
using namespace std;

// Structure of a Node
struct Node {
	int data;
	struct Node* link;
};

struct Queue {
	struct Node *front, *rear;
};

// Function to create Circular queue
void enQueue(Queue* q, int value)
{
	struct Node* temp = new Node;
	temp->data = value;
	if (q->front == NULL)
		q->front = temp;
	else
		q->rear->link = temp;

	q->rear = temp;
	q->rear->link = q->front;
}

// Function to delete element from Circular Queue
int deQueue(Queue* q)
{
	if (q->front == NULL) {
		cout << "Queue is empty";
		return INT_MIN;
	}

	// If this is the last node to be deleted
	int value; // Value to be dequeued
	if (q->front == q->rear) {
		value = q->front->data;
		free(q->front);
		q->front = NULL;
		q->rear = NULL;
	}
	else // There are more than one nodes
	{
		struct Node* temp = q->front;
		value = temp->data;
		q->front = q->front->link;
		q->rear->link = q->front;
		free(temp);
	}

	return value;
}

// Function displaying the elements of Circular Queue
void displayQueue(struct Queue* q)
{
	struct Node* temp = q->front;
	cout << endl << "Elements in Circular Queue are: ";
	while (temp->link != q->front) {
		cout << temp->data << " ";
		temp = temp->link;
	}
	cout << temp->data;
}

/* Driver of the program */
int main()
{
	// Create a queue and initialize front and rear
	Queue* q = new Queue;
	q->front = q->rear = NULL;

	// Inserting elements in Circular Queue
	enQueue(q, 14);
	enQueue(q, 22);
	enQueue(q, 6);

	// Display elements present in Circular Queue
	displayQueue(q);

	// Deleting elements from Circular Queue
	cout << endl << "Deleted value = " << deQueue(q);
	cout << endl << "Deleted value = " << deQueue(q);

	// Remaining elements in Circular Queue
	displayQueue(q);

	enQueue(q, 9);
	enQueue(q, 20);
	displayQueue(q);

	return 0;
}
